sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/GroupByKeyEncodingTest.java

```diff
@@ -0,0 +1,221 @@
+      key1 = k1;
+      key2 = k2;
+      perKey = pk;
+      val = v;
+    }
+  }
+
+  static class MakeKey extends SimpleFunction<DumbData, Key> {
+
+    @Override
+    public Key apply(DumbData dd) {
+      return new Key(dd.key1, dd.key2);
+    }
+  }
+
+  static class Generator extends DoFn<Config, DumbData> {
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      Config cfg = c.element();
+
+      ArrayList<String> bs = new ArrayList<>(cfg.key1);
+      for (int i = 0; i < cfg.key1; i++) {
+        bs.add("hereGoesLongStringID" + i);
+      }
+
+      for (int k = 0; k < cfg.perKey; k++) {
+        for (long j = 0; j < cfg.key2; j++) {
+          for (int i = 0; i < cfg.key1; i++) {
+            DumbData dd = new DumbData();
+            dd.key1 = bs.get(i);
+            dd.key2 = j;
+            dd.value1 = cfg.val;
+            c.output(dd);
+          }
+        }
+      }
+    }
+  }
+
+  static class MergeGbk extends DoFn<KV<Key, Iterable<DumbData>>, String> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MergeGbk.class);
+
+    private final Aggregator<Long, Long> keyCnt =
+        createAggregator("key count", new Sum.SumLongFn());
+
+    private final Aggregator<Long, Long> itemCnt =
+        createAggregator("item count", new Sum.SumLongFn());
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      Key key = c.element().getKey();
+      Iterable<DumbData> data = c.element().getValue();
+
+      keyCnt.addValue(1L);
+      long count = 0;
+      for (DumbData val : data) {
+        count++;
+      }
+      itemCnt.addValue(count);
+
+      c.output(key.key1 + "," + key.key2 + "," + count);
+      if (count == 0) {
+        LOG.info("no data for (" + key.key1 + "," + key.key2 + ")");
+      } else {
+        LOG.info(count + " data items for (" + key.key1 + "," + key.key2 + ")");
+      }
+    }
+  }
+
+  @Test
+  @Category(RunnableOnService.class)
+  public void testGroupByKeyWithBogusKey() throws Exception {
+    final int NUM_VALUES = 10;
+
+    Pipeline p = TestPipeline.create();
+
+    // Create several Configs, so that several parallel Generator instances emit
+    // data. Otherwise, a pre-shuffle combine would already combine everything
+    // and results would be correct even with a bogus key.hashCode().
+    PCollection<KV<Key, DumbData>> dataset1 = p
+        .apply(Create.of(
+            new Config(3, 5, NUM_VALUES, 2),
+            new Config(3, 5, NUM_VALUES, 2),
+            new Config(3, 5, NUM_VALUES, 2),
+            new Config(3, 5, NUM_VALUES, 2)))
+        .apply(ParDo.of(new Generator()))
+        .apply(WithKeys.of(new MakeKey()));
+
+    PCollection<String> result = dataset1
+        .apply(GroupByKey.<Key, DumbData>create())
+        .apply(ParDo.of(new MergeGbk()));
+
+    PAssert.that(result).containsInAnyOrder(
+        "hereGoesLongStringID0,0," + (NUM_VALUES * 4),
```

Comment:
This seems like an `Iterable` that it would be better to generate with a couple of simple for loops - it seems as though it would be much more readable.

You could alternatively (concurrently?) split this into two assertions; one that all keys are present as expected (preferably using Structural Values), and the other is that all keys have the expected number of occurrences.